test/e2e/multicast_test.go

```diff
@@ -158,7 +278,261 @@ type multicastTestcase struct {
+}
+
+// testMulticastStatsWithSendersReceivers tests multiple multicast senders and receivers cases with two AntreaNetworkPolicies applied. One anp blocks or allows IGMP traffic and another blocks or allows multicast traffic.
+// It checks results of all the multicast-related commands, including kubectl get multicastgroups, antctl get podmulticaststats and kubectl get antreanetworkpolicystats.
+func testMulticastStatsWithSendersReceivers(t *testing.T, data *TestData, mc multicastStatsTestcase) {
+	mcjoinWaitTimeout := defaultTimeout / time.Second
+	var igmpQueryType = int32(0x11)
+
+	for _, senderConfig := range mc.senderConfigs {
+		_, _, cleanupFunc := createAndWaitForPodWithExactName(t, data, data.createMcJoinPodOnNode, senderConfig.name, senderConfig.nodeName, data.testNamespace, false)
+		defer cleanupFunc()
+	}
+
+	for _, receiverConfig := range mc.receiverConfigs {
+		_, _, cleanupFunc := createAndWaitForPodWithExactName(t, data, data.createMcJoinPodOnNode, receiverConfig.name, receiverConfig.nodeName, data.testNamespace, false)
+		defer cleanupFunc()
+	}
+
+	var err error
+	k8sUtils, err = NewKubernetesUtils(data)
+	failOnError(err, t)
+	p10 := float64(10)
+	udp := v1.ProtocolUDP
+
+	for _, anp := range mc.multicastANPConfigs {
+		np := &crdv1alpha1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{Namespace: data.testNamespace, Name: anp.name, Labels: map[string]string{"antrea-e2e": anp.name}},
+			Spec: crdv1alpha1.NetworkPolicySpec{
+				Priority: p10,
+				AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+					{PodSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"antrea-e2e": anp.appliedToPod}}},
+				},
+				Egress: []crdv1alpha1.Rule{},
+			},
+		}
+		for i, _ := range anp.ruleConfigs {
+			np.Spec.Egress = append(np.Spec.Egress, crdv1alpha1.Rule{
+				To: []crdv1alpha1.NetworkPolicyPeer{
+					{
+						IPBlock: &crdv1alpha1.IPBlock{
+							CIDR: fmt.Sprintf("%s/32", anp.ruleConfigs[i].address),
+						},
+					},
+				},
+				Ports: []crdv1alpha1.NetworkPolicyPort{
+					{
+						Protocol: &udp,
+					},
+				},
+				Action: &anp.ruleConfigs[i].action,
+				Name:   anp.ruleConfigs[i].name,
+			})
+		}
+		if _, err = k8sUtils.CreateOrUpdateANP(np); err != nil {
+			failOnError(fmt.Errorf("create ANP failed for ANP %s: %v", np.Name, err), t)
+		}
+		failOnError(data.waitForANPRealized(t, data.testNamespace, np.Name), t)
+	}
+
+	for _, anp := range mc.igmpANPConfigs {
+		np := &crdv1alpha1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{Namespace: data.testNamespace, Name: anp.name, Labels: map[string]string{"antrea-e2e": anp.name}},
+			Spec: crdv1alpha1.NetworkPolicySpec{
+				Priority: p10,
+				AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+					{PodSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"antrea-e2e": anp.appliedToPod}}},
+				},
+				Egress:  []crdv1alpha1.Rule{},
+				Ingress: []crdv1alpha1.Rule{},
+			},
+		}
+		for i := range anp.ruleConfigs {
+			if anp.ruleConfigs[i].address == "224.0.0.1" {
```

Comment:
It is not a good idea to assume a specific IP address to decide the IGMP rule direction (IGMP type), you can add the property in the `ruleConfig`?

Reply:
updated